numpyro/contrib/funsor/enum_messenger.py

```diff
@@ -0,0 +1,634 @@
+
+    def __exit__(self, *args, **kwargs):
+        if self.keep:
+            # don't keep around references to other frames
+            old_frame = _DIM_STACK.pop()
+            saved_frame = StackFrame(
+                name_to_dim=old_frame.name_to_dim, dim_to_name=old_frame.dim_to_name,
+                parents=(), iter_parents=(), keep=self.keep
+            )
+            self._saved_frames.append(saved_frame)
+        else:
+            _DIM_STACK.pop()
+        return super().__exit__(*args, **kwargs)
+
+
+class GlobalNamedMessenger(NamedMessenger):
+
+    def __init__(self, fn):
+        self._saved_globals = ()
+        super().__init__(fn)
+
+    def __enter__(self):
+        if self._ref_count == 0:
+            for name, dim in self._saved_globals:
+                _DIM_STACK.global_frame.write(name, dim)
+            self._saved_globals = ()
+        return super().__enter__()
+
+    def __exit__(self, *args, **kwargs):
+        if self._ref_count == 1:
+            for name, dim in self._saved_globals:
+                _DIM_STACK.global_frame.free(name, dim)
+        return super().__exit__(*args, **kwargs)
+
+    def postprocess_message(self, msg):
+        if msg["type"] == "to_funsor":
+            self._pyro_post_to_funsor(msg)
+        elif msg["type"] == "to_data":
+            self._pyro_post_to_data(msg)
+
+    def _pyro_post_to_funsor(self, msg):
+        if msg["kwargs"]["dim_type"] in (DimType.GLOBAL, DimType.VISIBLE):
+            for name in msg["value"].inputs:
+                self._saved_globals += ((name, _DIM_STACK.global_frame.name_to_dim[name]),)
+
+    def _pyro_post_to_data(self, msg):
+        if msg["kwargs"]["dim_type"] in (DimType.GLOBAL, DimType.VISIBLE):
+            for name in msg["args"][0].inputs:
+                self._saved_globals += ((name, _DIM_STACK.global_frame.name_to_dim[name]),)
+
+
+class BaseEnumMessenger(NamedMessenger):
+    """
+    Handles first_available_dim management, enum effects should inherit from this
+    """
+    def __init__(self, fn, first_available_dim=None):
+        assert first_available_dim is None or first_available_dim < 0, first_available_dim
+        self.first_available_dim = first_available_dim
+        super().__init__(fn)
+
+    def __enter__(self):
+        if self._ref_count == 0 and self.first_available_dim is not None:
+            self._prev_first_dim = _DIM_STACK.set_first_available_dim(self.first_available_dim)
+        return super().__enter__()
+
+    def __exit__(self, *args, **kwargs):
+        if self._ref_count == 1 and self.first_available_dim is not None:
+            _DIM_STACK.set_first_available_dim(self._prev_first_dim)
+        return super().__exit__(*args, **kwargs)
+
+
+##########################################
+# User-facing handler implementations
+##########################################
+
+class plate(GlobalNamedMessenger):
+    """
+    An alternative implementation of :class:`numpyro.primitives.plate` primitive.
```

Comment:
Clarify here that only this version of `plate` is compatible with enumeration?  Might also want to add a pointer to the `plate_to_enum_plate` context manager.